Python/ceval.c

```diff
@@ -3417,6 +3417,34 @@ _PyEval_LoadName(PyThreadState *tstate, _PyInterpreterFrame *frame, PyObject *na
     return value;
 }
 
+_PyStackRef
+_PyForIter_NextWithIndex(PyObject *seq, _PyStackRef index)
+{
+    assert(PyStackRef_IsTaggedInt(index));
+    assert(PyTuple_CheckExact(seq) || PyList_CheckExact(seq));
+    intptr_t i = PyStackRef_UntagInt(index);
+    if (PyTuple_CheckExact(seq)) {
+        size_t size = PyTuple_GET_SIZE(seq);
+        if ((size_t)i >= size) {
+            return PyStackRef_NULL;
+        }
+        return PyStackRef_FromPyObjectNew(PyTuple_GET_ITEM(seq, i));
+    }
+    size_t size = PyList_GET_SIZE(seq);
+    if ((size_t)i >= size) {
+        return PyStackRef_NULL;
+    }
+#ifdef Py_GIL_DISABLED
+    PyObject *item = _PyList_GetItemRef((PyListObject *)seq, i);
+    if (item == NULL) {
+        return PyStackRef_NULL;
+    }
+    return PyStackRef_FromPyObjectSteal(item);
+#else
+    return PyStackRef_FromPyObjectNew(PyList_GET_ITEM(seq, i));
+#endif
```

Comment:
``_PyList_GetItemRef`` does the same thing as this entire block (check index and get the item). It also behaves differently on FT and default builds, so you can just use that here, and replace all this code.